File: hub/src/OpenAudioLink.Core/Audio/RationalResampler.cs
namespace OpenAudioLink.Core.Audio;

/// <summary>
/// Streaming sample-rate conversion between two fixed rates.
///
/// Spotify is 44.1 kHz and the RTP profile is 48 kHz, so something has to
/// resample and docs/CAST-POINTS.md puts it at the Hub: one clock domain
/// for the whole house is worth more than the CPU it costs.
///
/// The ratio is exact — 44100:48000 reduces to 147:160 — so this is a
/// polyphase FIR rather than an interpolator chasing a drifting phase.
/// Conceptually the input is upsampled by <c>L</c>, low-pass filtered and
/// decimated by <c>M</c>; in practice only the taps that land on a nonzero
/// sample are ever multiplied, which is <see cref="TapsPerPhase"/> of them
/// per output sample.
///
/// Measured against an ideal sine, the worst error is about -110 dB, the
/// response is flat to 20 kHz and images sit below -90 dB. That is far
/// better than the lossy source it will carry, which is the point: the
/// resampler should not be the thing anyone can hear.
///
/// Not thread-safe: it carries the filter history, so one instance belongs
/// to one stream.
/// </summary>
public sealed class RationalResampler
{
    /// <summary>
    /// Taps per phase. Sets the transition band: 64 puts the passband edge
    /// above 20 kHz for 44.1 kHz input, where 32 would pull it down to
    /// about 18 kHz. The cost is 64 multiplies per output sample, which at
    /// 48 kHz stereo is a few million a second — nothing.
    /// </summary>
    public const int TapsPerPhase = 64;

    /// <summary>
    /// Kaiser window parameter, chosen for roughly -90 dB stopband.
    /// </summary>
    private const double WindowBeta = 8.6;

    private readonly int _channels;
    private readonly int _interpolation;
    private readonly int _decimation;

    /// <summary>Phase-major: <c>_taps[p * TapsPerPhase + k]</c>.</summary>
    private readonly float[] _taps;

    /// <summary>Circular history, one slot per frame, newest at <see cref="_newest"/>.</summary>
    private readonly float[] _history;

    private int _newest;

    /// <summary>
    /// Position of the next output within the current input frame,
    /// numerator-only: an output is due while this is below
    /// <see cref="_interpolation"/>.
    /// </summary>
    private int _phase;

    public RationalResampler(int inputRate, int outputRate, int channels)
    {
        if (inputRate <= 0 || outputRate <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(inputRate), "Sample rates must be positive.");
        }
        if (channels <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(channels), "Channel count must be positive.");
        }

        int divisor = GreatestCommonDivisor(inputRate, outputRate);
        _interpolation = outputRate / divisor;
        _decimation = inputRate / divisor;
        _channels = channels;

        InputRate = inputRate;
        OutputRate = outputRate;

        _taps = BuildTaps(_interpolation, _decimation);
        _history = new float[TapsPerPhase * channels];
    }

    public int InputRate { get; }

    public int OutputRate { get; }

    public int Channels => _channels;

    /// <summary>
    /// True when input and output rates are equal, so callers can skip the
    /// whole thing rather than filter audio that needs no filtering.
    /// </summary>
    public bool IsPassthrough => _interpolation == 1 && _decimation == 1;

    /// <summary>
    /// Group delay in output frames. The filter is symmetric, so this is
    /// exactly half its length: <see cref="TapsPerPhase"/> / 2 input frames.
    /// </summary>
    public double DelayFrames => TapsPerPhase / 2.0 * OutputRate / InputRate;

    /// <summary>
    /// Largest number of output frames <paramref name="inputFrames"/> can
    /// produce. The count varies from call to call because the ratio is not
    /// an integer, so a caller sizing a buffer must use this rather than the
    /// average — the very first frame of a 44.1 to 48 kHz conversion
    /// already produces two.
    /// </summary>
    public int MaxOutputFrames(int inputFrames) =>
        inputFrames <= 0
            ? 0
            // Worst case is starting at phase zero and ending just short of
            // the next input: ((frames + 1) * L - 1) / M.
            : (int)((((long)inputFrames + 1) * _interpolation - 1) / _decimation);

    /// <summary>
    /// Resamples interleaved frames.
    /// </summary>
    /// <param name="input">Interleaved input, a whole number of frames.</param>
    /// <param name="output">
    /// Interleaved output, at least <see cref="MaxOutputFrames"/> long.
    /// </param>
    /// <returns>Samples written, which is frames times channels.</returns>
    public int Process(ReadOnlySpan<float> input, Span<float> output)
    {
        int frames = input.Length / _channels;
        if (frames == 0)
        {
            return 0;
        }

        if (output.Length < MaxOutputFrames(frames) * _channels)
        {
            throw new ArgumentException(
                $"Output needs room for {MaxOutputFrames(frames)} frames.", nameof(output));
        }

        int written = 0;
        for (int frame = 0; frame < frames; frame++)
        {
            _newest = _newest + 1 == TapsPerPhase ? 0 : _newest + 1;
            int slot = _newest * _channels;
            for (int channel = 0; channel < _channels; channel++)
            {
                _history[slot + channel] = input[frame * _channels + channel];
            }

            while (_phase < _interpolation)
            {
                int phaseAt = _phase * TapsPerPhase;
                for (int channel = 0; channel < _channels; channel++)
                {
                    double sum = 0.0;
                    int at = _newest;
                    for (int k = 0; k < TapsPerPhase; k++)
                    {
                        sum += _taps[phaseAt + k] * _history[at * _channels + channel];
                        at = at == 0 ? TapsPerPhase - 1 : at - 1;
                    }
                    output[written + channel] = (float)sum;
                }
                written += _channels;
                _phase += _decimation;
            }
            _phase -= _interpolation;
        }

        return written;
    }

    /// <summary>
    /// Forgets the filter history. Called when a stream restarts, so the
    /// tail of the last track cannot bleed into the first samples of the
    /// next one.
    /// </summary>
    public void Reset()
    {
        Array.Clear(_history);
        _newest = 0;
        _phase = 0;
    }

    /// <summary>
    /// Windowed-sinc low-pass, cut at the lower of the two Nyquist limits
    /// and scaled by the interpolation factor to make up for the zeros
    /// inserted between input samples.
    /// </summary>
    private static float[] BuildTaps(int interpolation, int decimation)
    {
        int length = TapsPerPhase * interpolation;
        double cutoff = 0.5 / Math.Max(interpolation, decimation);
        double centre = length / 2.0;
        double normaliser = BesselI0(WindowBeta);

        var prototype = new double[length];
        for (int i = 0; i < length; i++)
        {
            double x = i - centre;
            double sinc = Math.Abs(x) < 1e-9
                ? 2.0 * cutoff
                : Math.Sin(2.0 * Math.PI * cutoff * x) / (Math.PI * x);

            double position = x / centre;
            double window = BesselI0(WindowBeta * Math.Sqrt(Math.Max(0.0, 1.0 - position * position)))
                / normaliser;

            prototype[i] = sinc * window * interpolation;
        }

        // Stored phase-major so one output sample walks a contiguous run.
        var taps = new float[length];
        for (int phase = 0; phase < interpolation; phase++)
        {
            for (int k = 0; k < TapsPerPhase; k++)
            {
                taps[phase * TapsPerPhase + k] = (float)prototype[phase + k * interpolation];
            }
        }
        return taps;
    }

    /// <summary>Modified Bessel function of the first kind, order zero.</summary>
    private static double BesselI0(double x)
    {
        double sum = 1.0;
        double term = 1.0;
        for (int k = 1; k < 64; k++)
        {
            double factor = x / (2.0 * k);
            term *= factor * factor;
            sum += term;
            if (term < 1e-14 * sum)
            {
                break;
            }
        }
        return sum;
    }

    private static int GreatestCommonDivisor(int a, int b)
    {
        while (b != 0)
        {
            (a, b) = (b, a % b);
        }
        return a;
    }
}

File: phone/core/src/main/kotlin/org/openaudiolink/core/RationalResampler.kt
package org.openaudiolink.core

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Streaming sample-rate conversion between two fixed rates.
 *
 * Spotify is 44.1 kHz and the wire is 48 kHz (decision 13), so something
 * has to resample. On the Hub that is `RationalResampler.cs`; this is the
 * same filter, ported, because a phone producing a different 48 kHz from
 * the Hub's would be a difference nobody could hear but everybody would
 * have to reason about.
 *
 * The ratio is exact — 44100:48000 reduces to 147:160 — so this is a
 * polyphase FIR rather than an interpolator chasing a drifting phase.
 * Conceptually the input is upsampled by `L`, low-pass filtered and
 * decimated by `M`; in practice only the taps landing on a nonzero sample
 * are multiplied, which is [TAPS_PER_PHASE] of them per output sample.
 *
 * Not thread-safe: it carries the filter history, so one instance belongs
 * to one stream.
 */
class RationalResampler(
    val inputRate: Int,
    val outputRate: Int,
    val channels: Int,
) {
    companion object {
        /**
         * Taps per phase. Sets the transition band: 64 puts the passband
         * edge above 20 kHz for 44.1 kHz input, where 32 would pull it down
         * to about 18 kHz. The cost is 64 multiplies per output sample,
         * which at 48 kHz stereo is a few million a second — nothing, even
         * on a phone.
         */
        const val TAPS_PER_PHASE = 64

        /** Kaiser window parameter, chosen for roughly -90 dB stopband. */
        private const val WINDOW_BETA = 8.6

        private fun gcd(a: Int, b: Int): Int {
            var x = a
            var y = b
            while (y != 0) {
                val t = x % y
                x = y
                y = t
            }
            return x
        }

        /** Modified Bessel function of the first kind, order zero. */
        private fun besselI0(x: Double): Double {
            var sum = 1.0
            var term = 1.0
            for (k in 1 until 64) {
                val factor = x / (2.0 * k)
                term *= factor * factor
                sum += term
                if (term < 1e-14 * sum) break
            }
            return sum
        }

        /**
         * Windowed-sinc low-pass, cut at the lower of the two Nyquist
         * limits and scaled by the interpolation factor to make up for the
         * zeros inserted between input samples.
         */
        private fun buildTaps(interpolation: Int, decimation: Int): FloatArray {
            val length = TAPS_PER_PHASE * interpolation
            val cutoff = 0.5 / max(interpolation, decimation)
            val centre = length / 2.0
            val normaliser = besselI0(WINDOW_BETA)

            val prototype = DoubleArray(length)
            for (i in 0 until length) {
                val x = i - centre
                val sinc = if (abs(x) < 1e-9) {
                    2.0 * cutoff
                } else {
                    sin(2.0 * PI * cutoff * x) / (PI * x)
                }
                val position = x / centre
                val window = besselI0(
                    WINDOW_BETA * sqrt(max(0.0, 1.0 - position * position))
                ) / normaliser
                prototype[i] = sinc * window * interpolation
            }

            // Stored phase-major so one output sample walks a contiguous run.
            val taps = FloatArray(length)
            for (phase in 0 until interpolation) {
                for (k in 0 until TAPS_PER_PHASE) {
                    taps[phase * TAPS_PER_PHASE + k] =
                        prototype[phase + k * interpolation].toFloat()
                }
            }
            return taps
        }
    }

    private val interpolation: Int
    private val decimation: Int
    private val taps: FloatArray
    private val history: FloatArray
    private var newest = 0

    /**
     * Position of the next output within the current input frame,
     * numerator-only: an output is due while this is below [interpolation].
     */
    private var phase = 0

    init {
        require(inputRate > 0 && outputRate > 0) { "sample rates must be positive" }
        require(channels > 0) { "channel count must be positive" }

        val divisor = gcd(inputRate, outputRate)
        interpolation = outputRate / divisor
        decimation = inputRate / divisor
        taps = buildTaps(interpolation, decimation)
        history = FloatArray(TAPS_PER_PHASE * channels)
    }

    /** True when the rates match, so a caller can skip the whole thing. */
    val isPassthrough: Boolean get() = interpolation == 1 && decimation == 1

    /**
     * Largest number of output frames [inputFrames] can produce.
     *
     * The count varies from call to call because the ratio is not an
     * integer, so a caller sizing a buffer must use this rather than the
     * average — the very first frame of a 44.1 to 48 kHz conversion already
     * produces two.
     */
    fun maxOutputFrames(inputFrames: Int): Int =
        if (inputFrames <= 0) 0
        else (((inputFrames.toLong() + 1) * interpolation - 1) / decimation).toInt()

    /**
     * Resamples interleaved frames.
     *
     * @return samples written, which is frames times channels.
     */
    fun process(input: FloatArray, inputLength: Int, output: FloatArray): Int {
        val frames = inputLength / channels
        if (frames == 0) return 0
        require(output.size >= maxOutputFrames(frames) * channels) {
            "output needs room for ${maxOutputFrames(frames)} frames"
        }

        var written = 0
        for (frame in 0 until frames) {
            newest = if (newest + 1 == TAPS_PER_PHASE) 0 else newest + 1
            val slot = newest * channels
            for (channel in 0 until channels) {
                history[slot + channel] = input[frame * channels + channel]
            }

            while (phase < interpolation) {
                val phaseAt = phase * TAPS_PER_PHASE
                for (channel in 0 until channels) {
                    var sum = 0.0
                    var at = newest
                    for (k in 0 until TAPS_PER_PHASE) {
                        sum += taps[phaseAt + k] * history[at * channels + channel]
                        at = if (at == 0) TAPS_PER_PHASE - 1 else at - 1
                    }
                    output[written + channel] = sum.toFloat()
                }
                written += channels
                phase += decimation
            }
            phase -= interpolation
        }

        return written
    }

    /**
     * Forgets the filter history.
     *
     * Called when a stream restarts, so the tail of the last track cannot
     * bleed into the first samples of the next one.
     */
    fun reset() {
        history.fill(0f)
        newest = 0
        phase = 0
    }
}
